File: 프로그래머스/연습문제/Lv3/연속 펄스 부분 수열의 합/연속 펄스 부분 수열의 합.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

long long solution(vector<int> sequence) {
    vector<long long> DP[2];    // DP[0][i]: x1 했을 때, DP[1][i]: x(-1) 했을 때
    DP[0].push_back(sequence[0]);
    DP[1].push_back(-sequence[0]);
    
    for(int i=1; i<sequence.size(); i++){
        DP[0].push_back(max(DP[1][i-1]+sequence[i], (long long)sequence[i]));
        DP[1].push_back(max(DP[0][i-1]-sequence[i], (long long)-sequence[i]));
    }
    
    return max(*max_element(DP[0].begin(), DP[0].end()), *max_element(DP[1].begin(), DP[1].end()));
}

File: 프로그래머스/연습문제/Lv3/연속 펄스 부분 수열의 합/연속 펄스 부분 수열의 합.py
def solution(sequence):
    DP = [[0 for i in range(len(sequence))] for i in range(2)]
    DP[0][0] = sequence[0]
    DP[1][0] = -sequence[0]

    for i in range(1, len(sequence)):
        DP[0][i] = max(DP[1][i-1]+sequence[i], sequence[i])
        DP[1][i] = max(DP[0][i-1]-sequence[i], -sequence[i])

    return max(max(DP[0]), max(DP[1]))
